File: src/main/kotlin/com/io/reactivecoroutines/weather/WeatherSearchHandler.java
package com.io.reactivecoroutines.weather;

import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;
import reactor.util.Logger;
import reactor.util.Loggers;

import java.time.LocalDate;
import java.util.List;
import java.util.Random;
import java.util.stream.IntStream;

@Component
public class WeatherSearchHandler {

    private static final Logger LOGGER = Loggers.getLogger(WeatherSearchHandler.class);
    private final WeatherService weatherService;

    public WeatherSearchHandler(WeatherService weatherService) {
        this.weatherService = weatherService;
    }

    Mono<ServerResponse> searchByExample(ServerRequest req) {
        return req.bodyToMono(WeatherOptionFilters.class)
                .flatMap(filters -> ServerResponse
                        .ok()
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(weatherService.queryByExample(
                                new WeatherInfo(
                                        filters.region,
                                        filters.country,
                                        filters.state,
                                        filters.city,
                                        filters.localDate,
                                        filters.avgTemperature
                                )
                        ), WeatherInfo.class)
                        .doOnNext(a -> LOGGER.debug("searchByExample onNext"))
                );
    }

    Mono<ServerResponse> searchOneByExample(ServerRequest req) {
        return req.bodyToMono(WeatherOptionFilters.class)
                .flatMap(filters -> ServerResponse
                        .ok()
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(
                                weatherService.queryOneByExample(
                                        new WeatherInfo(
                                                filters.region,
                                                filters.country,
                                                filters.state,
                                                filters.city,
                                                filters.localDate,
                                                filters.avgTemperature
                                        )
                                ), WeatherInfo.class)
                        .doOnNext(a -> LOGGER.debug("searchOneByExample onNext"))
                );
    }

    Mono<ServerResponse> searchRandomOneByExample(ServerRequest ignoredReq) {
        // bridge some blocking code here with a scheduler
        return Mono.fromCallable(this::getRandomCriteria)
                .subscribeOn(Schedulers.boundedElastic())
                .flatMap(randomDateCriteria -> ServerResponse
                        .ok()
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(weatherService.queryOneByExample(randomDateCriteria), WeatherInfo.class)
                        .doOnNext(a -> LOGGER.debug("searchRandomOneByExample onNext"))
                        .doOnSuccess(a -> LOGGER.debug("searchRandomOneByExample doOnSuccess"))
                );
    }

    private WeatherInfo getRandomCriteria() {
        // simulate a blocking implementation
        try {
            LOGGER.debug("getRandomCriteria");
            Thread.sleep(10);
            LOGGER.debug("getRandomCriteria after sleep");
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }

        List<Integer> daysList = IntStream.rangeClosed(1, 30).boxed().toList();
        int randomDayOfMonth = daysList.get(new Random().nextInt(daysList.size()));

        return new WeatherInfo(
                "Europe",
                "Finland",
                "",
                "Helsinki",
                LocalDate.of(2019, 7, randomDayOfMonth),
                null
        );
    }

    private record WeatherOptionFilters(
            String region,
            String country,
            String state,
            String city,
            LocalDate localDate,
            String avgTemperature
    ) {
    }
}

File: src/main/kotlin/com/io/reactivecoroutines/weather/WeatherSearchHandler.kt
package com.io.reactivecoroutines.weather

import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.flatMapConcat
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.reactive.awaitFirst
import org.springframework.http.MediaType
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.server.ServerRequest
import org.springframework.web.reactive.function.server.ServerResponse
import org.springframework.web.reactive.function.server.bodyAndAwait
import org.springframework.web.reactive.function.server.bodyValueAndAwait
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import reactor.util.Logger
import reactor.util.Loggers
import java.time.LocalDate

private val log: Logger = Loggers.getLogger(WeatherSearchHandler::class.java)

@Component
class WeatherSearchHandler(
    private val weatherService: WeatherService
) {

    suspend fun searchByExample(req: ServerRequest): ServerResponse {

        val filters = req.bodyToMono(WeatherOptionFilters::class.java).awaitFirst()

        return ServerResponse
            .ok()
            .contentType(MediaType.APPLICATION_JSON)
            .bodyValueAndAwait(
                weatherService.queryByExample(
                    WeatherInfo(
                        id = null,
                        region = filters.region,
                        country = filters.country,
                        state = filters.state,
                        city = filters.city,
                        localDate = filters.localDate,
                        avgTemperature = filters.avgTemperature
                    )
                )
            )
    }

    suspend fun searchOneByExample(req: ServerRequest): ServerResponse {

        val filters = req.bodyToMono(WeatherOptionFilters::class.java).awaitFirst()

        return ServerResponse
            .ok()
            .contentType(MediaType.APPLICATION_JSON)
            .bodyValueAndAwait(
                weatherService.queryOneByExample(
                    WeatherInfo(
                        id = null,
                        region = filters.region,
                        country = filters.country,
                        state = filters.state,
                        city = filters.city,
                        localDate = filters.localDate,
                        avgTemperature = filters.avgTemperature
                    )
                ).takeIf {
                    it.isNotEmpty()
                }
                    ?: throw IllegalArgumentException("No weather data found for filters: $filters")
            )
    }

    suspend fun searchRandomOneByExample(ignoredReq: ServerRequest): ServerResponse {
        // bridge some blocking code here with a scheduler
        val randomCriteria = Mono.fromCallable(this::getRandomCriteria)
            .subscribeOn(Schedulers.boundedElastic())

        return ServerResponse
            .ok()
            .contentType(MediaType.APPLICATION_JSON)
            .bodyValueAndAwait(
                weatherService.queryOneByExample(randomCriteria.awaitFirst())
            ).also {
                log.debug("searchRandomOneByExample doOnSuccess")
            }

    }

    suspend fun getWeatherStream(request: ServerRequest): ServerResponse {
        return ServerResponse.ok().bodyAndAwait(
            (1..7666L)
                .asFlow()
                .onEach { delay(100) }
                .flatMapConcat { daysToAdd ->
                    weatherService.queryOneByExample(
                        WeatherInfo(
                            id = null,
                            region = "Europe",
                            country = "Denmark",
                            state = "",
                            city = "Copenhagen",
                            localDate = LocalDate.of(1995, 1, 1).plusDays(daysToAdd),
                            avgTemperature = null
                        )
                    ).asFlow()
                }
        )
    }

    private fun getRandomCriteria(): WeatherInfo {
        // simulate a blocking implementation
        try {
            log.debug("getRandomCriteria")
            Thread.sleep(100)
            log.debug("getRandomCriteria after sleep")
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }

        val randomDayOfMonth: Int = (1..30).random()

        return WeatherInfo(
            region = "Europe",
            country = "Finland",
            state = "",
            city = "Helsinki",
            localDate = LocalDate.of(2019, 7, randomDayOfMonth),
            avgTemperature = null
        )
    }

    private data class WeatherOptionFilters(
        val region: String?,
        val country: String?,
        val state: String?,
        val city: String?,
        val localDate: LocalDate?,
        val avgTemperature: String?
    )
}
